Add RegisterPreCompilationSourceOutput API for incremental generators (#83088)
dotnet/roslyn: docs/features/pre-compilation-source-outputs.md

Reviewer: Let's update the cookbook with this API and usage examples.

Author: Done

Critically, this improvement comes without adding any new compilation phases to the driver. The pre-compilation outputs are simply additional steps that run before the compilation is finalized - the total number of *compilation* objects created by the driver remains the same as today (the initial compilation is augmented in-place with `AddSyntaxTrees`, not rebuilt from scratch).

### Additional Scenarios

**Cross-Generator Type Visibility**: As a secondary benefit, pre-compilation sources are visible to *all* generators' standard phases, not just the generator that produced them. A generator that reads `.proto` files and emits C# types via `RegisterPreCompilationSourceOutput` makes those types available to other generators that perform binding in their `RegisterSourceOutput` phase. This enables cross-generator interoperability without requiring generators to know about each other.

> **Configurable Post-Initialization**: Generator authors want to emit attributes or other marker types during post-initialization, but need to read `AnalyzerConfigOptions` or additional files to determine *which* sources to emit (see [#53632](https://github.com/dotnet/roslyn/issues/53632)). `RegisterPostInitializationOutput` cannot accept any inputs, forcing authors to either emit all possible sources unconditionally or split their generator into multiple packages. `RegisterPreCompilationSourceOutput` addresses this directly: the generated source is added to the initial compilation just like post-init source, but the generator can read parse options, analyzer config options, and additional files to decide what to emit.

Reviewer 2: Should this issue (#53632) be marked as closed by the PR? Also once merged we should make sure people see this under the feature requests for source generator ordering.

    /// </summary>
    /// <remarks>
    /// The source value provider must not depend on compilation or syntax node inputs.
    /// Attempting to access the compilation or syntax trees during this phase will throw
    /// an <see cref="InvalidOperationException"/>.
    /// </remarks>
    [Experimental(RoslynExperiments.PreCompilationSourceOutput)]
    public void RegisterPreCompilationSourceOutput<TSource>(
        IncrementalValueProvider<TSource> source,
        Action<PreCompilationSourceProductionContext, TSource> action);

    [Experimental(RoslynExperiments.PreCompilationSourceOutput)]
    public void RegisterPreCompilationSourceOutput<TSource>(
        IncrementalValuesProvider<TSource> source,
        Action<PreCompilationSourceProductionContext, TSource> action);
}
```

The production context is a new type, distinct from `SourceProductionContext`:

```csharp
[Experimental(RoslynExperiments.PreCompilationSourceOutput)]
public readonly struct PreCompilationSourceProductionContext
{
    public CancellationToken CancellationToken { get; }

    public void AddSource(string hintName, string source);
    public void AddSource(string hintName, SourceText sourceText);
}
```

> `PreCompilationSourceProductionContext` intentionally does **not** include `ReportDiagnostic`. Pre-compilation is an early phase focused purely on producing source; diagnostic reporting should be done in the standard phase.

Reviewer: ```suggestion
`PreCompilationSourceProductionContext` intentionally does **not** include `ReportDiagnostic`. Pre-compilation is an early phase focused purely on producing source; diagnostic reporting should be done in a separate analyzer.
```

A new `IncrementalGeneratorOutputKind` value is also introduced:

```csharp
[Flags]
public enum IncrementalGeneratorOutputKind
{
    // ... existing values ...

    /// <summary>
    /// A pre-compilation source output, registered via
    /// <see cref="IncrementalGeneratorInitializationContext.RegisterPreCompilationSourceOutput"/>.
    /// Source is added to the initial compilation before compilation-dependent phases execute.
    /// </summary>
    PreCompilation = 0b10000
}
```

> ### Usage Example

Reviewer: This, plus some explanation, can likely go directly into the cookbook.

Author: Done

```csharp
// This compiles fine, but will fail at runtime!
context.RegisterPreCompilationSourceOutput(
    context.CompilationProvider,  // <- compilation is not available in this phase
    (ctx, compilation) =>
    {
        // This code never executes - an exception is thrown when evaluating the provider
    });
```

To catch this, the `DriverStateTable.Builder` does **not** have the `Compilation` or `SyntaxStore` set during the pre-compilation phase. Accessing either property throws an `InvalidOperationException`. This exception is thrown from the `DriverStateTable.Builder` property getter, caught by `InputNode.UpdateStateTable` (which wraps it in a `UserFunctionException`), and handled by the pre-compilation error handler in the driver.

When a pre-compilation output fails (whether from accessing compilation-dependent inputs or from any other exception), the generator is placed in **error state**: a diagnostic is reported, and the generator's standard phase is **skipped entirely**. Other generators are unaffected - their pre-compilation and standard phases continue to execute normally.

### Why Not Compile-Time Safety?

> A more principled approach would introduce a parallel type system (`PreCompilationValueProvider<T>` / `PreCompilationValuesProvider<T>`) that prevents compilation-dependent providers from being passed to `RegisterPreCompilationSourceOutput` at compile time. However, this would require duplicating every transform operation (`Select`, `Where`, `Collect`, `SelectMany`, `Combine`, `WithComparer`, `WithTrackingName`) for the pre-compilation provider types, creating a significantly larger API surface. Given that this is an experimental API, runtime enforcement with clear error messages is sufficient for the initial release. If the feature proves valuable and the runtime safety issue becomes a real pain point, a compile-time safe type system can be pursued in a future iteration.

Reviewer 2: Or perhaps we could have an analyzer for this.

Author: Unfortunately you can't accurately identify the inputs for a graph statically, you have to execute the graph itself, because nodes could be added by eg. other methods (this happens in razor). So that essentially boils down the to halting problem :(

We could have a simple case analyzer, but I'm not sure how much value that would bring because its probably in the complex cases that you aren't going to easily see the error without running it.

Author: Tracked by #83333 -- @333fred asked that we ship at least a best-effort analyzer for the statically-obvious cases, even though a fully accurate one isn't possible (see my earlier reply for why).

## Incremental Step Tracking

Pre-compilation outputs have a distinct step name: `WellKnownGeneratorOutputs.PreCompilationSourceOutput` (value `"PreCompilationSourceOutput"`). This is separate from `SourceOutput` and `ImplementationSourceOutput`, allowing tools to distinguish pre-compilation steps in incremental step tracking output.

Per-generator `GeneratorRunStateTable.Builder` instances are shared across both the pre-compilation and standard passes, so all steps for a generator (both pre-compilation and standard) appear in a unified view.

Pre-compilation sources appear in `GeneratorRunResult.GeneratedSources` alongside post-init and standard sources.

## Parse Options Reparse

> Like post-initialization trees, pre-compilation trees are reparsed when parse options change between driver runs. A unified `RequiresConstantTreeReparse` check handles both post-init and pre-compilation trees in a single pass.

Reviewer 2: Cannot pre-compilation phases depend on parse options and produce entirely different trees, so simple reparsing wouldn't be enough?

Author: Yes, that's a good point! I'll fix and add a test to check that scenario.

Author: Ah right, we will always re-run anything that relies on `ParseOptions` so it'll still be correct. This check is for any trees that *dont* depend on `ParseOptions`, so wouldn't normally re-run, but we need to reparse them with the new options. 

In theory we could only apply this to trees that didn't rely on the parse options, but that requires tracking. Currently we only do tracking when explicitly enabled as it slows the pipeline. We could probably enable some subset for these trees, but I'm not sure it's worth it: we can revisit if we see numbers in traces suggesting we're double parsing lots of things on parse options change.
